runtime/parachains/src/scheduler.rs

```diff
@@ -13,3 +13,1403 @@
+				Some(parachains[core_index.0 as usize])
+			}
+			Some(CoreOccupied::Parathread(ref entry)) => Some(entry.claim.0),
+		}
+	}
+
+	/// Get the validators in the given group, if the group index is valid for this session.
+	#[allow(unused)]
+	pub(crate) fn group_validators(group_index: GroupIndex) -> Option<Vec<ValidatorIndex>> {
+		ValidatorGroups::get().get(group_index.0 as usize).map(|g| g.clone())
+	}
+
+	/// Get the group assigned to a specific core by index at the current block number. Result undefined if the core index is unknown
+	/// or the block number is less than the session start index.
+	pub(crate) fn group_assigned_to_core(core: CoreIndex, at: T::BlockNumber) -> Option<GroupIndex> {
+		let config = <configuration::Module<T>>::config();
+		let session_start_block = <SessionStartBlock<T>>::get();
+
+		if at < session_start_block { return None }
+
+		if config.parachain_rotation_frequency.is_zero() {
+			// interpret this as "no rotations"
+			return Some(GroupIndex(core.0));
+		}
+
+		let validator_groups = ValidatorGroups::get();
+
+		if core.0 as usize >= validator_groups.len() { return None }
+
+		let rotations_since_session_start: T::BlockNumber =
+			(at - session_start_block) / config.parachain_rotation_frequency.into();
+
+		let rotations_since_session_start
+			= match <T::BlockNumber as TryInto<u32>>::try_into(rotations_since_session_start)
+		{
+			Ok(i) => i,
+			Err(_) => 0, // can only happen if rotations occur only once every u32::max(),
+			             // so functionally no difference in behavior.
+		};
+
+		let group_idx = (core.0 as usize + rotations_since_session_start as usize) % validator_groups.len();
+		Some(GroupIndex(group_idx as u32))
+	}
+
+	/// Returns an optional predicate that should be used for timing out occupied cores.
+	///
+	/// If `None`, no timing-out should be done. The predicate accepts the index of the core, and the
+	/// block number since the last validator group rotation, and the respective parachain and parathread
+	/// timeouts, i.e. only within `max(config.chain_availability_period, config.thread_availability_period)`
+	/// of the last rotation would this return `Some`.
+	#[allow(unused)]
+	pub(crate) fn availability_timeout_predicate() -> Option<impl Fn(CoreIndex, T::BlockNumber) -> bool> {
+		let now = <system::Module<T>>::block_number();
+		let config = <configuration::Module<T>>::config();
+
+		let session_start = <SessionStartBlock<T>>::get();
+		let blocks_since_session_start = now.saturating_sub(session_start);
+		let blocks_since_last_rotation = blocks_since_session_start % config.parachain_rotation_frequency;
+
+		let absolute_cutoff = sp_std::cmp::max(
+			config.chain_availability_period,
+			config.thread_availability_period,
+		);
+
+		let availability_cores = AvailabilityCores::get();
+
+		if blocks_since_last_rotation >= absolute_cutoff {
+			None
+		} else {
+			Some(move |core_index: CoreIndex, pending_since| {
+				match availability_cores.get(core_index.0 as usize) {
```

Comment:
```suggestion
				match availability_cores.get(core_index.0 as usize).flatten() {
```
less explicit, but merges the both `None` and `Some(None)` cases, and simplifies the match

Reply:
Yeah, I thought about that after seeing `flatten` in your PR, but I decided to be more explicit on this one.